new persistence api

raskolnikoov opened the pull request:
Rewrite of persistence api and added ravendb persistence plugin. Maybe ravendb/couchebase should be moved outside ofta the core framework into its own repos?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/AsynkronIT/protoactor-dotnet?pullRequest=89) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/AsynkronIT/protoactor-dotnet?pullRequest=89) before we can accept your contribution.<br/><hr/>**Daniel Söderberg** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/AsynkronIT/protoactor-dotnet?pullRequest=89) it.</sub>

raskolnikoov commented:
Im not 100% sure I got the versioning right. Byt maybe that is not set in the projects.

rogeralsing commented:
Can you describe the changes to start a discussion about it?
How is snapshotting handled now?

raskolnikoov commented:
I have removed the "GetSnapshotInterval" method and property from Persistence. I also removed the bool property "Recovering". I think snapshots should be triggered by the actor itself instead when snapshots are needed.

I've introduced the following messages to be send to the actor on recovery/persisting:

- RecoveryStarted
- RecoveryCompleted
- RecoverSnapshot *contains persisted snapshot object
- RecoverEvent *contains persisted event object
- PersistedEvent *contains the persisted event object

raskolnikoov commented:
"IProviderState" has been expanded with methods to handle "DeleteEventsAsync" and "DeleteSnapshotsAsync" so the actor could clean up stored events/snapshots that are not needed anymore.

cpx86 commented:
Letting the actor control snapshotting makes sense IMO. It is more related to the needs of the actor than the needs of the persistence.

About versioning, don't worry. It's currently set by the build script and we only bump it after merging to master.

raskolnikoov commented:
Anyone time ro review my changes?

rogeralsing commented:
The change looks good to me.
cc @cpx86

cpx86 commented on src/Proto.Persistence/InMemoryProviderState.cs:
```diff
@@ -60,5 +51,15 @@ public Task PersistSnapshotAsync(string actorName, ulong eventIndex, object snap
+
+        public Task DeleteEventsAsync(string actorName, ulong fromEventIndex)
+        {
+            throw new NotImplementedException();
```
We should just return `Task.FromResult(0)` here.

cpx86 commented on src/Proto.Persistence/InMemoryProviderState.cs:
```diff
@@ -60,5 +51,15 @@ public Task PersistSnapshotAsync(string actorName, ulong eventIndex, object snap
+
+        public Task DeleteSnapshotsAsync(string actorName, ulong fromEventIndex)
+        {
+            throw new NotImplementedException();
```
These should probably be implemented as well. If we want to keep this class - I'm not sure if it's needed for anything except testing (in which case it should be moved)

raskolnikoov replied:
I think we should remove InMemoryProvider. Dont see the use case for it really. If we need a default provider maybe we should persist om disk in a File.

cpx86 replied:
Do it then :)

cpx86 commented:
Looks good! I like the introduction of the recovery messages, so the actor can handle those differently if necessary :)

cpx86 commented:
Just fix the merge conflict and I'll merge 👍

rogeralsing commented:
Merged, thanks!
